icicle/src/merkle-tree/mmcs.cu

```diff
@@ -0,0 +1,458 @@
+    cudaStream_t& stream = tree_config.ctx.stream;
+
+    if (number_of_inputs == 0) THROW_ICICLE_ERR(IcicleError_t::InvalidArgument, "No matrices provided");
+
+    if (compression.preimage_max_length < tree_config.arity * tree_config.digest_elements)
+      THROW_ICICLE_ERR(
+        IcicleError_t::InvalidArgument,
+        "Hash max preimage length does not match merkle tree arity multiplied by digest elements");
+
+    std::vector<Matrix<L>> sorted_inputs(number_of_inputs);
+    std::partial_sort_copy(
+      inputs, inputs + number_of_inputs, sorted_inputs.begin(), sorted_inputs.end(),
+      [](const Matrix<L>& left, const Matrix<L>& right) { return left.height > right.height; });
+
+    // Check that the height of any two given matrices either rounds up
+    // to the same next power of two or otherwise equal
+    for (unsigned int i = 0; i < number_of_inputs - 1; i++) {
+      unsigned int left = sorted_inputs[i].height;
+      unsigned int right = sorted_inputs[i + 1].height;
+
+      if (next_pow_of_two(left) == next_pow_of_two(right) && left != right)
+        THROW_ICICLE_ERR(
+          IcicleError_t::InvalidArgument, "Matrix heights that round up to the same power of two must be equal");
+    }
+
+    // for (int i = 0; i < number_of_inputs; i++) {
+    //   std::cout << "Orig Matrix " << i << ": " << inputs[i].width << ", " << inputs[i].height << std::endl;
+    //   for (int j = 0; j < inputs[i].height; j++) {
+    //     for (int k = 0; k < inputs[i].width; k++) {
+    //       std::cout << inputs[i].values[j * inputs[i].width + k] << " ";
+    //     }
+    //     std::cout << std::endl;
+    //   }
+    //   std::cout << "Matrix " << i << ": " << sorted_inputs[i].width << ", " << sorted_inputs[i].height << std::endl;
+    // }
+
+    uint64_t max_height = sorted_inputs[0].height;
+
+    // Calculate maximum additional memory needed for injected matrices
+    uint64_t max_aux_total_elements = 0;
+    uint64_t current_aux_total_elements = 0;
+    uint64_t current_height = 0;
+    uint64_t bottom_layer_leaves_elements = 0;
+    for (auto it = sorted_inputs.begin(); it < sorted_inputs.end(); it++) {
+      if (it->height == max_height) {
+        bottom_layer_leaves_elements += it->height * it->width;
+        continue;
+      }
+
+      if (it->height != current_height) {
+        current_height = it->height;
+        current_aux_total_elements = 0;
+      }
+
+      current_aux_total_elements += it->width * it->height;
+      if (current_aux_total_elements > max_aux_total_elements) { max_aux_total_elements = current_aux_total_elements; }
+    }
+
+    std::cout << "Max aux total elements " << max_aux_total_elements << std::endl;
+
+    uint64_t number_of_bottom_layer_rows = next_pow_of_two(max_height);
+    size_t leaves_info_memory = number_of_inputs * sizeof(Matrix<L>);
+
+    unsigned int tree_height = get_height(number_of_bottom_layer_rows);
+
+    // This will determine how much splitting do we need to do
```

Comment:
here too if you can make it a separate function it would be easier to understand what's going on here